### DISCUSSION TITLE
How to get recursive children with custom constraint?

### DISCUSSION THREAD

**Discussion by ilyapokrov** [staudenmeir/laravel-adjacency-list#245] [Q&A] [answered]
I have this structure in the Category model database: id, category_id, parent_id, title, total_products.
I want to get a tree of nested child categories where the ID of the top parent is 3.
In the category model I specified:
```
use \Staudenmeir\LaravelAdjacencyList\Eloquent\HasRecursiveRelationships;
public function getLocalKeyName()
   {
      return 'category_id';
   }
```
I do it this way and everything works:
`$tree = Category::find(3)->descendantsAndSelf;`
But I want to add a custom constraint - the number of total_products is not equal to 0.
I could not find such an example in the documentation where it would be possible to specify two things at the same time - the ID of the top parent and my constraint.
I tried to simply get the tree, without specifying the ID of the top parent:
```
$constraint = function ($query) {
   $query->where('total_products', '!=', '0');
};
$tree = Category::treeOf($constraint)->get();
```
But the result was not what I expected. The number of my categories has increased 5 times =))

How can I get the result of categories that have products in a tree?
```
[
  {
    "id": 1,
    "title": 'Cloth',
    "children": [
      {
        "id": 2,
        "title": 't-shirts and polos',
        "children": [
          {
            "id": 3,
            "title": 't-shirts',
            "children": []
          }
          {
            "id": 4,
            "title": 'polos',
            "children": []
          }
        ]
      },
      {
        "id": 5,
        "title": 'sweaters',
        "children": []
      }
    ]
  }
]
```
thank you in advance!

**Comment by staudenmeir** [comment]
Hi @ilyapokrov,
Do you mean like this?

```php
$tree = Category::find(3)->descendantsAndSelf()->where('total_products', '!=', '0')->get();
```

**Reply by ilyapokrov** [reply]
This is very similar to what I wanted, thank you)
What about the second question?
To get not just a collection, but a tree in this form:
```
[
  {
    "id": 1,
    "title": 'Cloth',
    "children": [
      {
        "id": 2,
        "title": 't-shirts and polos',
        "children": [
          {
            "id": 3,
            "title": 't-shirts',
            "children": []
          }
          {
            "id": 4,
            "title": 'polos',
            "children": []
          }
        ]
      },
      {
        "id": 5,
        "title": 'sweaters',
        "children": []
      }
    ]
  }
]
```
I tried doing this, but of course it doesn't work)
`$tree = Category::find(3)->descendantsAndSelf()->where('total_products', '!=', '0')->toTree()->get();`

**Reply by staudenmeir** [reply]
You need to apply `->toTree()` _after_ executing the query:

```php
$tree = Category::find(3)->descendantsAndSelf()->where('total_products', '!=', '0')->get()->toTree();
```

**Reply by ilyapokrov** [reply]
Strange, but it gives an error in this case.
`BadMethodCallException
Method Illuminate\Database\Eloquent\Collection::toTree does not exist.`

**Reply by staudenmeir** [reply]
What version of the package are you using?

**Reply by ilyapokrov** [reply]
I tried to install it like this, but there was an error:
`composer require staudenmeir/laravel-adjacency-list:"^1.0"`
I searched for this error in the questions and found the answer:
https://github.com/staudenmeir/laravel-adjacency-list/issues/53
So I installed this version:
`composer require staudenmeir/laravel-adjacency-list:"^1.6"`

**Reply by staudenmeir** [reply]
What version does `composer show` show for `staudenmeir/laravel-adjacency-list`?

**Reply by ilyapokrov** [reply]
staudenmeir/laravel-adjacency-list v1.6     Recursive Laravel Eloquent relationships with CTEs

**Reply by staudenmeir** [reply, marked as answer]
> composer require staudenmeir/laravel-adjacency-list:"^1.6"

The `toTree()` isn't available in `v1.6`, but your command should actually install `v1.9.6` because that's the latest version compatible with Laravel 8. What operating system and terminal are you using?

Try running `composer require staudenmeir/laravel-adjacency-list:"^1.9.6"` instead.

**Reply by ilyapokrov** [reply]
Thank you! This version works!
